packages/material-ui-utils/src/useId.ts

```diff
@@ -0,0 +1,18 @@
+import * as React from 'react';
+
+/**
+ * Private module reserved for @material-ui packages.
```

Comment:
Could we resolve what the intent of `@material-ui/utils` is? This is exported as `unstable_` but annotated as private. The current state of `@material-ui/utils` is a mixed bag that can not be understood reasonably by npm users and it looks like even we don't know what this package is.

We'd do us a big favor if we would just rename it to `/internal` so we don't have to worry about maintaining utils and then we bump the required version in the dependent packages on every release.

It's not a blocker for this PR but definitely for the next release.

Reply:
@eps1lon Agree! Let's try to come up with an answer. You are right unstable mean we plan to make it public while private, well, is private. However, I don't think that we should solve it here. For later.

Maybe the following?

- `@material-ui/utils` -> `@material-ui/internal`
- `@material-ui/types` -> `@material-ui/internal`
- `@material-ui/unstyled` depends on `@material-ui/internal`
- `@material-ui/system` depends on `@material-ui/internal`
- `@material-ui/styles` depends on `@material-ui/internal`
- `@material-ui/unstyled` reexport some of the utils we want to make public
- `@material-ui/core` reexport some of the utils we want to make public, from `@material-ui/unstyled`

Reply:
@eps1lon I have noticed `@material-ui/types` used in the documentation for one use case:

https://github.com/mui-org/material-ui/blob/0aefc928b00788b0ec5f74842b7c9704a354cf4a/docs/src/pages/components/breadcrumbs/RouterBreadcrumbs.tsx#L15

do we still need it?

Reply:
> do we still need it with the new versions of TypeScript?

We don't need it for new versions of typescript but haven't had any discussions about bumping the required versions.

Reply:
> reexport some of the utils we want to make public

Then what is the point of moving them into an internal package if we just re-export them? That's not what internal means.

Reply:
> haven't had any discussions about bumping the required versions.

@eps1lon Do we need a discussion about changing the supported version of TypeScript?

> Then what is the point of moving them into an internal package if we just re-export them

Great question, the only value I can think of is to allow `@material-ui/system` to depend on `@material-ui/internal` rather than `@material-ui/unstyled`.

Reply:
>  Do we need a discussion about changing the supported version of TypeScript?

Like any other breaking change: yes.

Reply:
> Like any other breaking change: yes.

Sorry, I wasn't clear. Do we want to change the minimum version of TypeScript supported? I'm personally not familiar with the matter.

Reply:
I haven't though about it but since we didn't change anything we still need an Omit polyfill.

Reply:
It's just that we test with the  latest version where we don't need it. But 3.2 does which is what we support.